crates/hir/src/lib.rs

```diff
@@ -1679,6 +1679,12 @@ impl DefWithBody {
             acc.extend(AnyDiagnostic::inference_diagnostic(db, self.into(), d, &source_map));
         }
         for (pat_or_expr, mismatch) in infer.type_mismatches() {
+            // Silence mismatches involving unresolved projections
+            let is_unknown_proj = |ty: &Ty| ty.contains_unknown() && ty.is_projection();
+            if is_unknown_proj(&mismatch.expected) || is_unknown_proj(&mismatch.actual) {
```

Comment:
What about types like `S<<T as Tr>::U>`?

Reply:
Hmm... It's a little complicated. Perhaps we need to recursively check whether it satisfies `has_unknown && is_projection`?

Reply:
I think so, see https://github.com/rust-lang/rust-analyzer/blob/fc0d51ae655dd17ffe73fdf0b9a4819d262a8e1b/crates/hir-ty/src/infer.rs#L698-L708 and https://github.com/rust-lang/rust-analyzer/blob/fc0d51ae655dd17ffe73fdf0b9a4819d262a8e1b/crates/hir-ty/src/infer.rs#L1647 :grimacing:.

Reply:
I've read the code of `chalk`, and the logic for recursive checking seems somewhat similar to the part where flags are computed here: [https://github.com/rust-lang/chalk/blob/12eda8042b764134509ec2c12883786b9674cccd/chalk-ir/src/lib.rs#L688](https://github.com/rust-lang/chalk/blob/12eda8042b764134509ec2c12883786b9674cccd/chalk-ir/src/lib.rs#L688).

Reply:
Thanks!